src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -265,6 +265,23 @@ public ValueTask<FlushResult> WriteDataAsync(int streamId, StreamOutputFlowContr
             }
         }
 
+        public void WriteData(int streamId, StreamOutputFlowControl flowControl, in ReadOnlySequence<byte> data, bool endStream)
+        {
+            // The Length property of a ReadOnlySequence can be expensive, so we cache the value.
+            var dataLength = data.Length;
+
+            lock (_writeLock)
+            {
+                if (_completed || flowControl.IsAborted)
+                {
+                    return;
+                }
+
+                flowControl.Advance((int)dataLength);
```

Comment:
You skipped the flow control check `dataLength > flowControl.Available`. This could cause a protocol violation.

You'd probably have to do that check in the caller so you could divert to an async code path.

Reply:
Done